File: code/java/SocialMedia.java
import java.util.*;
import java.lang.*;
import java.io.*;

class SocialMedia {
	/* All Social Media Sites
	 * -------------------------------------------------
	 */

	/* All Social Media Sites ~ Nice Names
	 * -------------------------------------------------
	 */

	public Map<String, String> GetSocialMediaSites_NiceNames() {
		Map<String, String> socialmediasites = new HashMap<String, String>(){{
			put("add.this", "AddThis");
			put("blogger", "Blogger");
			put("buffer", "Buffer");
			put("diaspora", "Diaspora");
			put("digg", "Digg");
			put("douban", "Douban");
			put("email", "EMail");
			put("evernote", "EverNote");
			put("getpocket", "Pocket");
			put("facebook", "FaceBook");
			put("flattr", "Flattr");
			put("flipboard", "FlipBoard");
			put("google.bookmarks", "GoogleBookmarks");
			put("instapaper", "InstaPaper");
			put("line.me", "Line.me");
			put("linkedin", "LinkedIn");
			put("livejournal", "LiveJournal");
			put("gmail", "GMail");
			put("hacker.news", "HackerNews");
			put("ok.ru", "OK.ru");
			put("pinterest", "Pinterest");
			put("qzone", "QZone");
			put("reddit", "Reddit");
			put("renren", "RenRen");
			put("skype", "Skype");
			put("sms", "SMS");
			put("surfingbird.ru", "SurfingBird.ru");
			put("telegram.me", "Telegram.me");
			put("threema", "Threema");
			put("tumblr", "Tumblr");
			put("twitter", "Twitter");
			put("vk", "VK");
			put("weibo", "Weibo");
			put("xing", "Xing");
			put("yahoo", "Yahoo");
		}};
		
		return socialmediasites;
	}
		
				/* Social Media Sites With Share Links
				* -------------------------------------------------
				*/

	public String[] GetSocialMediaSites_WithShareLinks_OrderedByPopularity() {
		return new String[]{
			"google.bookmarks",
			"facebook",
			"reddit",
			"twitter",
			"linkedin",
			"tumblr",
			"pinterest",
			"blogger",
			"livejournal",
			"evernote",
			"add.this",
			"getpocket",
			"hacker.news",
			"digg",
			"buffer",
			"flipboard",
			"instapaper",
			"surfingbird.ru",
			"flattr",
			"diaspora",
			"qzone",
			"vk",
			"weibo",
			"ok.ru",
			"douban",
			"xing",
			"renren",
			"threema",
			"sms",
			"line.me",
			"skype",
			"telegram.me",
			"email",
			"gmail",
			"yahoo",
		};
	}
	
	public String[] GetSocialMediaSites_WithShareLinks_OrderedByAlphabet() {
		Map<String, String> socialmediaurls = GetSocialMediaSites_NiceNames();
		String[] socialmediasites = new String[socialmediaurls.size()];
		
		int index = 0;
		
		for(Map.Entry<String, String> entry : socialmediaurls.entrySet()) {
			String key = entry.getKey();
			socialmediasites[index++] = key;
		}

		Arrays.sort(socialmediasites);
		
		return socialmediasites;
	}
	
				/* Social Media Site Links With Share Links
				* -------------------------------------------------
				*/
	
	public Map<String, String> GetSocialMediaSiteLinks_WithShareLinks(Map<String, String> args) {
		String[] validargs = {
			"url",
			"title",
			"image",
			"desc",
			"appid",
			"redirecturl",
			"via",
			"hashtags",
			"provider",
			"language",
			"userid",
			"category",
			"phonenumber",
			"emailaddress",
			"cemailaddress",
			"bccemailaddress",
		};
		
		for (String validarg : validargs) {
			if(args.get(validarg) == null) {
				args.put(validarg, "");
			}
		}

		String text = args.get("title");
		String desc = args.get("desc");

		if(desc != null) {
			text += "%20%3A%20";
			text += desc;
		}

		args.put("text", text);
		
		Map<String, String> socialmediasites = new HashMap<String, String>(){{
			put("add.this", "http://www.addthis.com/bookmark.php?url=" + args.get("url"));
			put("blogger", "https://www.blogger.com/blog-this.g?u=" + args.get("url") + "&n=" + args.get("title") + "&t=" + args.get("desc"));
			put("buffer", "https://buffer.com/add?text=" + args.get("text") + "&url=" + args.get("url"));
			put("diaspora", "https://share.diasporafoundation.org/?title=" + args.get("title") + "&url=" + args.get("url"));
			put("digg", "http://digg.com/submit?url=" + args.get("url") + "&title=" + args.get("text"));
			put("douban", "http://www.douban.com/recommend/?url=" + args.get("url") + "&title=" + args.get("text"));
			put("email", "mailto:" + args.get("email_address") + "?subject=" + args.get("title") + "&body=" + args.get("desc"));
			put("evernote", "http://www.evernote.com/clip.action?url=" + args.get("url") + "&title=" + args.get("text"));
			put("getpocket", "https://getpocket.com/edit?url=" + args.get("url"));
			put("facebook", "http://www.facebook.com/sharer.php?u=" + args.get("url"));
			put("flattr", "https://flattr.com/submit/auto?user_id=" + args.get("user_id") + "&url=" + args.get("url") + "&title=" + args.get("title") + "&description=" + args.get("text") + "&language=" + args.get("language") + "&tags=" + args.get("hash_tags") + "&hidden=HIDDEN&category=" + args.get("category"));
			put("flipboard", "https://share.flipboard.com/bookmarklet/popout?v=2&title=" + args.get("text") + "&url=" + args.get("url"));
			put("gmail", "https://mail.google.com/mail/?view=cm&to=" + args.get("email_address") + "&su=" + args.get("title") + "&body=" + args.get("url") + "&bcc=" + args.get("bcc_email_address") + "&cc=" + args.get("cc_email_address"));
			put("google.bookmarks", "https://www.google.com/bookmarks/mark?op=edit&bkmk=" + args.get("url") + "&title=" + args.get("title") + "&annotation=" + args.get("text") + "&labels=" + args.get("hash_tags"));
			put("instapaper", "http://www.instapaper.com/edit?url=" + args.get("url") + "&title=" + args.get("title") + "&description=" + args.get("desc"));
			put("line.me", "https://lineit.line.me/share/ui?url=" + args.get("url") + "&text=" + args.get("text"));
			put("linkedin", "https://www.linkedin.com/shareArticle?mini=true&url=" + args.get("url") + "&title=" + args.get("title") + "&summary=" + args.get("text") + "&source=" + args.get("provider"));
			put("livejournal", "http://www.livejournal.com/update.bml?subject=" + args.get("text") + "&event=" + args.get("url"));
			put("hacker.news", "https://news.ycombinator.com/submitlink?u=" + args.get("url") + "&t=" + args.get("title"));
			put("ok.ru", "https://connect.ok.ru/dk?st.cmd=WidgetSharePreview&st.shareUrl=" + args.get("url"));
			put("pinterest", "http://pinterest.com/pin/create/button/?url=" + args.get("url"));
			put("qzone", "http://sns.qzone.qq.com/cgi-bin/qzshare/cgi_qzshare_onekey?url=" + args.get("url"));
			put("reddit", "https://reddit.com/submit?url=" + args.get("url") + "&title=" + args.get("title"));
			put("renren", "http://widget.renren.com/dialog/share?resourceUrl=" + args.get("url") + "&srcUrl=" + args.get("url") + "&title=" + args.get("text") + "&description=" + args.get("desc"));
			put("skype", "https://web.skype.com/share?url=" + args.get("url") + "&text=" + args.get("text"));
			put("sms", "sms:" + args.get("phone_number") + "?body=" + args.get("text"));
			put("surfingbird.ru", "http://surfingbird.ru/share?url=" + args.get("url") + "&description=" + args.get("desc") + "&screenshot=" + args.get("image") + "&title=" + args.get("title"));
			put("telegram.me", "https://t.me/share/url?url=" + args.get("url") + "&text=" + args.get("text") + "&to=" + args.get("phone_number"));
			put("threema", "threema://compose?text=" + args.get("text") + "&id=" + args.get("user_id"));
			put("tumblr", "https://www.tumblr.com/widgets/share/tool?canonicalUrl=" + args.get("url") + "&title=" + args.get("title") + "&caption=" + args.get("desc") + "&tags=" + args.get("hash_tags"));
			put("twitter", "https://twitter.com/intent/tweet?url=" + args.get("url") + "&text=" + args.get("text") + "&via=" + args.get("via") + "&hashtags=" + args.get("hash_tags"));
			put("vk", "http://vk.com/share.php?url=" + args.get("url") + "&title=" + args.get("title") + "&comment=" + args.get("desc"));
			put("weibo", "http://service.weibo.com/share/share.php?url=" + args.get("url") + "&appkey=&title=" + args.get("title") + "&pic=&ralateUid=");
			put("xing", "https://www.xing.com/spi/shares/new?url=" + args.get("url"));
			put("yahoo", "http://compose.mail.yahoo.com/?to=" + args.get("email_address") + "&subject=" + args.get("title") + "&body=" + args.get("text"));
		}};
		
		return socialmediasites;
	}
}

class main {
	public static void main (String[] args) {
		SocialMedia sm = new SocialMedia();
		Map<String, String> socialmediaargs = new HashMap<String, String>(){{
			put("url", "http://www.earthfluent.com/");
			put("title", "EarthFluent");
		}};
		Map<String, String> socialmediasites = sm.GetSocialMediaSiteLinks_WithShareLinks(socialmediaargs);
		
		//System.out.println(sm.GetSocialMediaSites_NiceNames().toString());
		
		//for (String socialmedia : sm.GetSocialMediaSites_WithShareLinks_OrderedByAlphabet()) {
		for (String socialmedia : sm.GetSocialMediaSites_WithShareLinks_OrderedByPopularity()) {
			System.out.println(socialmedia + " : " + socialmediasites.get(socialmedia));
		}
		//System.out.println(sm.GetSocialMediaSites_WithShareLinks_OrderedByPopularity());
	}
}

File: code/python/SocialMedia.py
class SocialMedia:
    def __init__(self):
        pass
    
    def GetSocialMediaSites_NiceNames(self):
        return {
            'add.this':'AddThis',
            'blogger':'Blogger',
            'buffer':'Buffer',
            'diaspora':'Diaspora',
            'digg':'Digg',
            'douban':'Douban',
            'email':'EMail',
            'evernote':'EverNote',
            'getpocket':'Pocket',
            'facebook':'FaceBook',
            'flattr':'Flattr',
            'flipboard':'FlipBoard',
            'google.bookmarks':'GoogleBookmarks',
            'instapaper':'InstaPaper',
            'line.me':'Line.me',
            'linkedin':'LinkedIn',
            'livejournal':'LiveJournal',
            'gmail':'GMail',
            'hacker.news':'HackerNews',
            'ok.ru':'OK.ru',
            'pinterest':'Pinterest',
            'qzone':'QZone',
            'reddit':'Reddit',
            'renren':'RenRen',
            'skype':'Skype',
            'sms':'SMS',
            'surfingbird.ru':'SurfingBird.ru',
            'telegram.me':'Telegram.me',
            'threema':'Threema',
            'tumblr':'Tumblr',
            'twitter':'Twitter',
            'vk':'VK',
            'weibo':'Weibo',
            'xing':'Xing',
            'yahoo':'Yahoo',
        }

    def GetSocialMediaSites_WithShareLinks_OrderedByPopularity(self):
        return [
            'google.bookmarks',
            'facebook',
            'reddit',
            'twitter',
            'linkedin',
            'tumblr',
            'pinterest',
            'blogger',
            'livejournal',
            'evernote',
            'add.this',
            'getpocket',
            'hacker.news',
            'digg',
            'buffer',
            'flipboard',
            'instapaper',
            'surfingbird.ru',
            'flattr',
            'diaspora',
            'qzone',
            'vk',
            'weibo',
            'ok.ru',
            'douban',
            'xing',
            'renren',
            'threema',
            'sms',
            'line.me',
            'skype',
            'telegram.me',
            'email',
            'gmail',
            'yahoo',
        ]
    
    def GetSocialMediaSites_WithShareLinks_OrderedByAlphabet(self):
        socialmediasites = self.GetSocialMediaSites_NiceNames().keys()
        socialmediasites.sort()
        return socialmediasites
    
    def GetSocialMediaSiteLinks_WithShareLinks(self, args):
        safeargs = [
            'url',
            'title',
            'image',
            'desc',
            'appid',
            'redirecturl',
            'via',
            'hash_tags',
            'provider',
            'language',
            'user_id',
            'category',
            'phone_number',
            'email_address',
            'cc_email_address',
            'bcc_email_address',
        ]
        
        for safearg in safeargs:
            if not args.get(safearg):
                args[safearg] = ''
        
        text = args['title']
        
        if len(args['desc']):
            text += '%20%3A%20' + args['desc']

        return {
            'add.this':'http://www.addthis.com/bookmark.php?url=' + args['url'],
            'blogger':'https://www.blogger.com/blog-this.g?u=' + args['url'] + '&n=' + args['title'] + '&t=' + args['desc'],
            'buffer':'https://buffer.com/add?text=' + text + '&url=' + args['url'],
            'diaspora':'https://share.diasporafoundation.org/?title=' + args['title'] + '&url=' + args['url'],
            'digg':'http://digg.com/submit?url=' + args['url'] + '&title=' + text,
            'douban':'http://www.douban.com/recommend/?url=' + args['url'] + '&title=' + text,
            'email':'mailto:' + args['email_address'] + '?subject=' + args['title'] + '&body=' + args['desc'],
            'evernote':'https://www.evernote.com/clip.action?url=' + args['url'] + '&title=' + text,
            'getpocket':'https://getpocket.com/edit?url=' + args['url'],
            'facebook':'http://www.facebook.com/sharer.php?u=' + args['url'],
            'flattr':'https://flattr.com/submit/auto?user_id=' + args['user_id'] + '&url=' + args['url'] + '&title=' + args['title'] + '&description=' + text + '&language=' + args['language'] + '&tags=' + args['hash_tags'] + '&hidden=HIDDEN&category=' + args['category'],
            'flipboard':'https://share.flipboard.com/bookmarklet/popout?v=2&title=' + text + '&url=' + args['url'], 
            'gmail':'https://mail.google.com/mail/?view=cm&to=' + args['email_address'] + '&su=' + args['title'] + '&body=' + args['url'] + '&bcc=' + args['bcc_email_address'] + '&cc=' + args['cc_email_address'],
            'google.bookmarks':'https://www.google.com/bookmarks/mark?op=edit&bkmk=' + args['url'] + '&title=' + args['title'] + '&annotation=' + text + '&labels=' + args['hash_tags'] + '',
            'instapaper':'http://www.instapaper.com/edit?url=' + args['url'] + '&title=' + args['title'] + '&description=' + args['desc'],
            'line.me':'https://lineit.line.me/share/ui?url=' + args['url'] + '&text=' + text,
            'linkedin':'https://www.linkedin.com/shareArticle?mini=true&url=' + args['url'] + '&title=' + args['title'] + '&summary=' + text + '&source=' + args['provider'],
            'livejournal':'http://www.livejournal.com/update.bml?subject=' + text + '&event=' + args['url'],
            'hacker.news':'https://news.ycombinator.com/submitlink?u=' + args['url'] + '&t=' + args['title'],
            'ok.ru':'https://connect.ok.ru/dk?st.cmd=WidgetSharePreview&st.shareUrl=' + args['url'],
            'pinterest':'http://pinterest.com/pin/create/button/?url=' + args['url'] ,
            'qzone':'http://sns.qzone.qq.com/cgi-bin/qzshare/cgi_qzshare_onekey?url=' + args['url'],
            'reddit':'https://reddit.com/submit?url=' + args['url'] + '&title=' + args['title'],
            'renren':'http://widget.renren.com/dialog/share?resourceUrl=' + args['url'] + '&srcUrl=' + args['url'] + '&title=' + text + '&description=' + args['desc'],
            'skype':'https://web.skype.com/share?url=' + args['url'] + '&text=' + text,
            'sms':'sms:' + args['phone_number'] + '?body=' + text,
            'surfingbird.ru':'http://surfingbird.ru/share?url=' + args['url'] + '&description=' + args['desc'] + '&screenshot=' + args['image'] + '&title=' + args['title'],
            'telegram.me':'https://t.me/share/url?url=' + args['url'] + '&text=' + text + '&to=' + args['phone_number'],
            'threema':'threema://compose?text=' + text + '&id=' + args['user_id'],
            'tumblr':'https://www.tumblr.com/widgets/share/tool?canonicalUrl=' + args['url'] + '&title=' + args['title'] + '&caption=' + args['desc'] + '&tags=' + args['hash_tags'],
            'twitter':'https://twitter.com/intent/tweet?url=' + args['url'] + '&text=' + text + '&via=' + args['via'] + '&hashtags=' + args['hash_tags'],
            'vk':'http://vk.com/share.php?url=' + args['url'] + '&title=' + args['title'] + '&comment=' + args['desc'],
            'weibo':'http://service.weibo.com/share/share.php?url=' + args['url'] + '&appkey=&title=' + args['title'] + '&pic=&ralateUid=',
            'xing':'https://www.xing.com/spi/shares/new?url=' + args['url'],
            'yahoo':'http://compose.mail.yahoo.com/?to=' + args['email_address'] + '&subject=' + args['title'] + '&body=' + text,
        }

sm = SocialMedia()
#socialmediasites = sm.GetSocialMediaSites_WithShareLinks_OrderedByAlphabet()
socialmediasites = sm.GetSocialMediaSites_WithShareLinks_OrderedByPopularity()
socialmediaurls = sm.GetSocialMediaSiteLinks_WithShareLinks({
    'url':'http://www.earthfluent.com/',
    'title':'EarthFluent',
})
#print(socialmediaurls.keys())

for socialmediasite in socialmediasites:
    print(socialmediasite + " : " + socialmediaurls[socialmediasite])
    pass
